### PR TITLE
bpo-24905: Support BLOB incremental I/O in sqlite module

### PR DESCRIPTION
**Pull request by palaviv**
This PR adds support in BLOB incremental I/O at the sqlite module. As asked by @serhiy-storchaka and @berkerpeksag I will try to get some more developers to give their input on the wanted API. I am tagging some people that are active in the ghaering/pysqlite and rogerbinns/apsw.
@ghaering, @rianhunter, @rogerbinns, @phdru. Please look at the PR and give your notes.

https://bugs.python.org/issue24905

### CODE REVIEW

**Comment by rogerbinns** [comment]
The APSW doc for reference is at https://rogerbinns.github.io/apsw/blob.html

Does having __len__ make sense?  Files don't have that method.  It is also confusing - should len return the size from the current seek offset?

The documentation should make clearer that you cannot change the size of a blob, and mention zeroblob as the means to make a blob in a query without having to fill it in.

It may be worth mentioning that another approach is to store large data in a file, and only store the filename in the database.  (This comes up on the sqlite-users mailing list quite a lot.)

**Comment by phdru** [comment]
I cannot remember I ever was in need to read/write a part of a BLOB; it was always "all or nothing" for me. So I never used BLOB APIs; instead I always SELECT/INSERT/UPDATE BLOB columns; in Postgres they are not even BLOB columns — I always use BYTEA type.

So I'm -0 on exposing BLOB API for SQLite.

**Comment by rogerbinns** [comment]
@phdru SQLite is the same with regular queries: you can only read or write blobs in their entirety.  That for example means that if you store a 25MB blob then you must read or write 25MB at once.

SQLite has the "incremental blob" API for accessing just portions of blobs.  The motivation comes from "Lite" in the name - developers use SQLite because it is lighter weight (amongst other reasons).  DBAPI doesn't specify incremental blob I/O so only developers intending to use SQLite directly and not another database would use it.  Should they be able to?

**Comment by phdru** [comment]
-0 from me means: I don't care and if there will be such an API I'm not gonna use it. That's all.

**Comment by palaviv** [comment]
Thanks for the input @rogerbinns.

> Does having len make sense? Files don't have that method. It is also confusing - should len return the size from the current seek offset?

What is the difference between implementing `__len__` to the method `length` APSW blob has?

**Comment by rogerbinns** [comment]
@palaviv there is no difference between the _value_ returned by __len__ and length or similar methods.  It is however very uncommon to have a __len__ method on file like objects - I couldn't find an example of any!  For example StringIO is closest and has no __len__.  Hence my recommendation to avoid __len__ in favour of another method name.

**Comment by serhiy-storchaka** [comment]
The `mmap.mmap()` object is an example of file-like object supporting `len()`.

**Comment by rogerbinns** [comment]
@serhiy-storchaka good example.  They don't [document](https://docs.python.org/3/library/mmap.html?highlight=mmap.mmap#mmap.mmapl) it though, and there is a size() method although it is returning something slightly different.  There also seems to be a correlation between types that have __len__ and those that can you can array access.

In any event my recommendation is to avoid breaking new ground with a __len__ method since that seems not to be normal practise for this kind of thing that provides a file like interface.

**Comment by palaviv** [comment]
I actually think that we should use `__len__` as by the [definition](https://docs.python.org/3/reference/datamodel.html#object.__len__) this is the length of the object. The `Blob` object is a representation of the BLOB and that is the BLOB length.

**Comment by serhiy-storchaka** [comment]
Pull request conversation is purposed for discussing the code.

It would be better to continue the design discussion on the bug tracker or mailing list.

**Comment by palaviv** [comment]
@serhiy-storchaka I have implemented the sequence protocol but I have a few questions:
1. Do I need both `PySequenceMethods.sq_item`, `PySequenceMethods.sq_ass_item` and `PyMappingMethods.mp_subscript`, `PyMappingMethods.mp_ass_subscript`.
2. I can't make `__contains__` work. Could you point me to how fix that?

**Comment by palaviv** [comment]
I think that the `contains` operation should not be supported for blobs. As blobs can be very large looking for a subset of bytes inside them will be a very inefficient process in memory or in compute.

File: Doc/library/sqlite3.rst
Comment (auvipy) [inline Doc/library/sqlite3.rst]: 3.8

File: Doc/library/sqlite3.rst
Comment (auvipy) [inline Doc/library/sqlite3.rst]: 3.8

File: Modules/_sqlite/blob.c
Comment (ZackerySpytz) [inline Modules/_sqlite/blob.c:228]: It seems that `data_buffer` is leaked here.

Comment (palaviv) [reply Modules/_sqlite/blob.c:228]: Hi @ZackerySpytz. I agree that there is a leak here.
This patch exist since January 2016 and I kind of given up on it ever going into CPython as there is no core developer that works on sqlite. I would recommend you to use apsw that support this feature. In case any core developer would be in interested in working on this I would gladly fix any needed changes.

**Comment by matrixise** [comment]
Hi @palaviv

Would you be interested to upgrade your PR to the last master?

Thank you

File: Modules/_sqlite/blob.c
Comment (palaviv) [reply Modules/_sqlite/blob.c:228]: Fix @ZackerySpytz. Lets hope that this will get in :)

File: Doc/library/sqlite3.rst
Comment (palaviv) [reply Doc/library/sqlite3.rst]: Done

File: Doc/library/sqlite3.rst
Comment (palaviv) [reply Doc/library/sqlite3.rst]: Done

File: Doc/library/sqlite3.rst
Comment (auvipy) [reply Doc/library/sqlite3.rst]: you can retarget this for py 3.9

File: Doc/library/sqlite3.rst
Comment (auvipy) [inline Doc/library/sqlite3.rst]: 3.9

**Review by auvipy — CHANGES_REQUESTED** [review]
Retarget for python 3.9

**Comment by eamanu** [comment]
Hi @palaviv There is some plan with this PR?
Was open 3 years ago. Are you still interest on this patch?
If yes, could you fix the conflict. please?

**Comment by palaviv** [comment]
Hi @eamanu,
This patch exist since January 2016 and I kind of given up on it ever going into CPython as there is no core developer that works on sqlite. I would recommend you to use apsw that support this feature. In case any core developer would be in interested in working on this I would gladly fix any needed changes.

**Comment by simonw** [comment]
I'd love to see this land in Python. I think there's a strong case for it: SQLite lets you store up to 2GB of data in a BLOB, and reading an entire 2GB value into memory at once isn't nearly as pleasant as reading it incrementally, which is what this would let us do.

**Comment by berkerpeksag** [comment]
I'm on my stay-at-home summer vacation and I'll try to get this landed (hopefully) this week.

**Comment by palaviv** [comment]
Thanks @berkerpeksag!!! I will work on a rebase now

**Comment by palaviv** [comment]
@berkerpeksag I have rebased and updated the documentation. Please let me know if there is anything more needed.

File: Modules/_sqlite/connection.c
Comment (berkerpeksag) [inline Modules/_sqlite/connection.c]: Do we need the second `NULL`?

File: Doc/includes/sqlite3/blob.py
Comment (berkerpeksag) [inline Doc/includes/sqlite3/blob.py]: It would be nice to make at least `readonly` a keyword-only argument.

File: Doc/includes/sqlite3/blob.py
Comment (berkerpeksag) [inline Doc/includes/sqlite3/blob.py]: It would be nice to change the data to something like:

```py
blob.write(b'Hello')
blob.write(b'World')
```

Comment (berkerpeksag) [reply Doc/includes/sqlite3/blob.py]: Also, it would be nice to expand the example to include other members of the API.

File: Doc/library/sqlite3.rst
Comment (berkerpeksag) [inline Doc/library/sqlite3.rst:886]: Nitpick: `Blob.*` prefix can be removed.

File: Doc/library/sqlite3.rst
Comment (berkerpeksag) [inline Doc/library/sqlite3.rst:894]: It would be nice to move this information into the Blob description. Something like:

```rst
[...] or by doing ``obj[:5]``. You can call ``len(obj)`` to get size of the BLOB.
```

File: Doc/library/sqlite3.rst
Comment (berkerpeksag) [inline Doc/library/sqlite3.rst]: Nitpick: "For example You" -> "For example, you"

File: Doc/library/sqlite3.rst
Comment (berkerpeksag) [inline Doc/library/sqlite3.rst]: Nitpick:

```rst
.. method:: Blob.seek(offset, whence=os.SEEK_SET)
```

File: Misc/NEWS.d/next/Library/2018-04-18-16-15-55.bpo-24905.jYqjYx.rst
Comment (berkerpeksag) [inline Misc/NEWS.d/next/Library/2018-04-18-16-15-55.bpo-24905.jYqjYx.rst]: We are using "Patch by Aviv Palivoda." format in NEWS entries.

File: Modules/_sqlite/blob.c
Comment (berkerpeksag) [inline Modules/_sqlite/blob.c]: I'm sure this is going to be an issue on the wild. It would be nice to fix this before merge.

File: Modules/_sqlite/blob.c
Comment (berkerpeksag) [inline Modules/_sqlite/blob.c]: Any reason we cannot use `PyErr_SetString` here?

File: Modules/_sqlite/blob.c
Comment (berkerpeksag) [inline Modules/_sqlite/blob.c]: Any reason we cannot use `PyErr_SetString` here?

File: Modules/_sqlite/blob.c
Comment (berkerpeksag) [inline Modules/_sqlite/blob.c]: Any reason we cannot use `PyErr_SetString` here?

File: Modules/_sqlite/blob.c
Comment (berkerpeksag) [inline Modules/_sqlite/blob.c]: Typo: cotains -> contains

File: Modules/_sqlite/blob.c
Comment (berkerpeksag) [inline Modules/_sqlite/blob.c:324]: I wonder if `PyExc_SystemError` is the best type of choice here and in other unsupported operations. I'll need to do some research on this.

File: Modules/_sqlite/blob.c
Comment (berkerpeksag) [inline Modules/_sqlite/blob.c:407]: It would be better if we can have more descriptive TODO comments. Perhaps we can help you resolve it before merge.

File: Modules/_sqlite/blob.c
Comment (berkerpeksag) [inline Modules/_sqlite/blob.c]: Nitpick: It would be better if we can use `{}` in new code.

File: Modules/_sqlite/blob.c
Comment (berkerpeksag) [inline Modules/_sqlite/blob.c]: Nitpick: Copy and paste error? :)

File: setup.py
Comment (berkerpeksag) [inline setup.py]: Nitpick: Can we add trailing comma here to avoid future noisy diffs?

```py
sqlite_srcs = [
    ...
    '_sqlite/util.c',
    '_sqlite/blob.c'
]
```

File: Modules/_sqlite/connection.h
Comment (berkerpeksag) [inline Modules/_sqlite/connection.h]: Nitpick: "statements, blobs and cursors" -> "statements, cursors and blobs"

File: Misc/NEWS.d/next/Library/2018-04-18-16-15-55.bpo-24905.jYqjYx.rst
Comment (berkerpeksag) [reply Misc/NEWS.d/next/Library/2018-04-18-16-15-55.bpo-24905.jYqjYx.rst]: It would be nice to add a whatsnew entry. We can use the following attribute format there:

```rst
(Contributed by Aviv Palivoda in :issue:`24905`.)
```

File: Lib/sqlite3/test/dbapi.py
Comment (berkerpeksag) [inline Lib/sqlite3/test/dbapi.py]: Nitpick: It would be better to keep the existing hanging indentation format here.

**Review by berkerpeksag — CHANGES_REQUESTED** [review]

File: Misc/NEWS.d/next/Library/2018-04-18-16-15-55.bpo-24905.jYqjYx.rst
Comment (palaviv) [reply Misc/NEWS.d/next/Library/2018-04-18-16-15-55.bpo-24905.jYqjYx.rst]: Should I add the whatsnew entry?

File: Modules/_sqlite/connection.c
Comment (palaviv) [reply Modules/_sqlite/connection.c]: 2 NULLs are better then none? Fixed :wink:

File: Modules/_sqlite/blob.c
Comment (palaviv) [reply Modules/_sqlite/blob.c:407]: That is a great advice as I do not remember what is that TODO for... I will try and figure this out but if I would not find anything I will remove it.

**Comment by palaviv** [comment]
Thanks for the review @berkerpeksag. I have made the requested changes; please review again.

**Comment by nightlark** [comment]
Other than rebasing (due new conflicts arising over time), is there anything that can be done to help move this PR along?

**Comment by erlend-aasland** [comment]
@nightlark, @palaviv: Here' a short list from the top-of-my head of what is needed to rebase this onto `main`:

- the test suite has been normalized; we now use snake case `test_foo_bar` method names
- Argument Clinic
- use heap types iso. static types
- exception types are accessed through the (temporary) global state; for Connection objects, it's available through `self->state`

**Comment by nightlark** [comment]
@erlend-aasland Okay — I think I understand how to use argument clinic. Is there a guide to what iso. static types (or heap types)? Is the iso. a prefix for the types or an abbreviation? If it’s an abbreviation maybe that’s the missing a search term I should be using to find relevant resources.

**Comment by erlend-aasland** [comment]
> @erlend-aasland Okay — I think I understand how to use argument clinic.

Great! It's nice once you get into it. Feel free to ask if you get stuck :)

> Is there a guide to what iso. static types (or heap types)?

There's some info in the [docs](https://docs.python.org/3/c-api/type.html#c.PyType_FromModuleAndSpec), but you can also check the PR's that converted the existing types:
- GH-22478
- GH-22444
- GH-22428
- GH-22417
- GH-26104, GH-26452 (see these PR's for proper GC support)

Don't hesitate to ask if you need more pointers.

> Is the iso. a prefix for the types or an abbreviation?

Sorry, it's an abbreviation: _instead of_ :) I have the bad habit of using it too much.

**Comment by erlend-aasland** [comment]
Regarding heap types, take a look at Victor's blog: https://vstinner.github.io/isolate-subinterpreters.html

**Comment by erlend-aasland** [comment]
@nightlark Are you still working on this PR?

**Comment by CoolCat467** [comment]
It's 2022 now woooo

**Comment by dholth** [comment]
+1

**Comment by JelleZijlstra** [comment]
I just merged #30680, a simplified version of this PR. Blobs will be in Python 3.11.

**Pull request closed without merging**
